java/monkey/src/main/java/root/Parser.java

```diff
@@ -0,0 +1,104 @@
+package root;
+
+import root.ast.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class Parser {
+
+    private final Lexer lexer;
+
+    private Token currentToken;
+
+    private Token peekToken;
+
+    public final List<String> errors = new ArrayList<>();
+
+    public Parser(Lexer lexer) {
+        this.lexer = lexer;
+
+        this.nextToken();
+        this.nextToken();
+    }
+
+    public Program parseProgram() {
+        Program program = new Program();
+
+        while (!curTokenIs(TokenType.EOF)) {
+            parseStatement().ifPresent(program.statements::add);
+            nextToken();;
+        }
+
+        return program;
+    }
+
+    private void peekError(TokenType expected) {
+        errors.add("Expected next token to be %s, got %s".formatted(expected.name(), currentToken.type().name()));
+    }
+
+    private void nextToken() {
+        this.currentToken = this.peekToken;
+        this.peekToken = this.lexer.nextToken();
+    }
+
+    private Optional<Statement> parseStatement() {
+        return switch (this.currentToken.type()) {
+            case LET -> this.parseLetStatement();
+            case RETURN -> this.parseReturnStatement();
+            default -> Optional.empty();
+        };
+    }
+
+    private Optional<Statement> parseReturnStatement() {
```

Comment:
I believe you're over using the `Optional` Class here.

We're always in control of the context, and we know that we'll either be dealing with a Stream of Tokens, or an end of Stream. I don't believe we need to wrap our logic in such a way.

Reply:
The main reason I'm return an optional in these functions is because they can fail and I rather return an Optional (since java doesn't have a native return type) then return null of throw an Exception the the case of an error.

Suppose someone writes `if (myVar < 10 { ` forgetting to close the bracket, it would result in an error when constructing the if expression, and I think return an Optional is the best case so we don't need multiple null checks or a try - catch

Reply:
Only in the context of library creation do we need to worry about someone else. As we control the entire context, we do not need to worry about that.

Pattern Matching (something we're showcasing a bit) can also handle nulls, so I'm not worried about returning null, or even a new `EmptyStatement`